File: public/codes/sorting/shellSort.js
async function shellSort(array, setArray, setComparingIndices, setSwappingIndices, setSortedIndices, speed, isSorting) {
    const n = array.length;

    for (let gap = Math.floor(n / 2); gap > 0; gap = Math.floor(gap / 2)) {
        for (let i = gap; i < n; i++) {
            let temp = array[i];
            let j;
            for (j = i; j >= gap && array[j - gap] > temp; j -= gap) {
                if (!isSorting.current) return;
                setComparingIndices([j, j - gap]);
                setArray([...array]);
                await sleep(speed);

                setSwappingIndices([j, j - gap]);
                array[j] = array[j - gap];
                setArray([...array]);
                await sleep(speed);
                setSwappingIndices([]);
            }
            array[j] = temp;
            setArray([...array]);
            await sleep(speed);
        }
    }

    setSortedIndices(Array.from({ length: n }, (_, i) => i));
    setComparingIndices([]);
}

// Helper function (assuming it's defined elsewhere or passed)
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

File: public/codes/sorting/shellSort.py
def shell_sort(array, set_array, set_comparing_indices, set_swapping_indices, set_sorted_indices, speed, is_sorting):
    n = len(array)

    gap = n // 2
    while gap > 0:
        for i in range(gap, n):
            temp = array[i]
            j = i
            while j >= gap and array[j - gap] > temp:
                if not is_sorting.current: return
                set_comparing_indices([j, j - gap])
                set_array(list(array))
                time.sleep(speed / 1000.0)

                set_swapping_indices([j, j - gap])
                array[j] = array[j - gap]
                set_array(list(array))
                time.sleep(speed / 1000.0)
                set_swapping_indices([])
                j -= gap
            array[j] = temp
            set_array(list(array))
            time.sleep(speed / 1000.0)
        gap //= 2

    set_sorted_indices(list(range(n)))
    set_comparing_indices([])

import time